group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,454 @@
+        }
+
+        if (request.generationIdOrMemberEpoch() < 0 && group.isInState(EMPTY)) {
+            // When the generation id is -1, the request comes from either the admin client
+            // or a consumer which does not use the group management facility. In this case,
+            // the request can commit offsets if the group is empty.
+            return;
+        }
+
+        Optional<String> groupInstanceId = OffsetCommitRequest.groupInstanceId(request);
+        if (request.generationIdOrMemberEpoch() >= 0 || !request.memberId().isEmpty() || groupInstanceId.isPresent()) {
+            // We are validating three things:
+            // 1. If the `groupInstanceId` is present, then it exists and is mapped to `memberId`;
+            // 2. The `memberId` exists in the group; and
+            // 3. The `generationId` matches the current generation id.
+            if (groupInstanceId.isPresent()) {
+                String memberId = group.staticMemberId(groupInstanceId.get());
+                if (memberId == null) {
+                    throw Errors.UNKNOWN_MEMBER_ID.exception();
+                } else if (!request.memberId().equals(memberId)) {
+                    throw Errors.FENCED_INSTANCE_ID.exception();
+                }
+            }
+
+            if (!group.hasMemberId(request.memberId())) {
+                throw Errors.UNKNOWN_MEMBER_ID.exception();
+            }
+
+            if (request.generationIdOrMemberEpoch() != group.generationId()) {
+                throw Errors.ILLEGAL_GENERATION.exception();
+            }
+        } else if (!group.isInState(EMPTY)) {
+            // If the request does not contain the member id and the generation
+            // id (version 0), offset commits are only accepted when the group
+            // is empty.
+            throw Errors.UNKNOWN_MEMBER_ID.exception();
+        }
+
+        if (group.isInState(COMPLETING_REBALANCE)) {
+            // We should not receive a commit request if the group has not completed rebalance;
+            // but since the consumer's member.id and generation is valid, it means it has received
+            // the latest group generation information from the JoinResponse.
+            // So let's return a REBALANCE_IN_PROGRESS to let consumer handle it gracefully.
+            throw Errors.REBALANCE_IN_PROGRESS.exception();
+        }
+    }
+
+    /**
+     * Validates an OffsetCommit request for a consumer group.
+     *
+     * @param group     The consumer group.
+     * @param context   The request context.
+     * @param request   The actual request.
+     */
+    public void validateOffsetCommitForConsumerGroup(
+        ConsumerGroup group,
+        RequestContext context,
+        OffsetCommitRequestData request
+    ) throws KafkaException {
+        if (request.generationIdOrMemberEpoch() < 0 && group.members().isEmpty()) {
+            // When the generation id is -1, the request comes from either the admin client
```

Comment:
i'm a bit confused here. generationIdOrMemberEpoch refers to the generation id if using the old protocol and member epoch when using the new protocol. from https://github.com/apache/kafka/pull/14046#discussion_r1270401089 we're assuming that a given offset commit request for a consumer group is in fact for the new protocol. 

then should the comment be `When the member epoch is -1`?

Reply:
Indeed, the comment is wrong.